### PR TITLE
tls config

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: http/client.go [hunk 1]
```diff
@@ -220,6 +221,48 @@ func (c *Client) Auth(username, password string) *Client {
...
 }
 
+type RoundTripperSetting struct {
+	CA   string
+	Cert string
+	Key  string
+}
+
+func (t *RoundTripperSetting) ToTLSRoundTripperSettings() promConfig.TLSRoundTripperSettings {
+	var output promConfig.TLSRoundTripperSettings
```
Comment (moshloop) [inline http/client.go:231]: I don't think we should use the prometheus imports - copy and paste instead

File: deps/deps.go [hunk 2]
```diff
@@ -202,7 +202,7 @@ var dependencies = map[string]Dependency{
 		Template: "https://storage.googleapis.com/kubebuilder-tools/kubebuilder-tools-{{.version}}-{{.os}}-{{.platform}}.tar.gz",
 	},
 	"postgrest": {
-		Version:    "v9.0.0.20211220",
```
Comment (adityathebe) [inline deps/deps.go:205]:
```
-- FAIL: TestInstallDependency (14.70s)
    --- FAIL: TestInstallDependency/postgrest (0.08s)
        deps_test.go:22: Installing postgrest
        deps_test.go:25: Failed to download postgrest: failed to download postgrest: failed to download https://github.com/PostgREST/postgrest/releases/download/v9.0.0.20211220/postgrest-v9.0.0.20211220-linux-static-x64.tar.xz: 404 Not Found
```

File: http/client.go [hunk 3]
```diff
@@ -220,6 +259,11 @@ func (c *Client) Auth(username, password string) *Client {
 	return c
 }
 
+type TLSConfig struct {
+	CA   string
+	Cert string
+}
+
```
Comment (moshloop) [inline http/client.go:266]:
```suggestion
type TLSConfig struct {
        // PEM encoded certificate of  of the CA to verify the server certificate 
	CA string `yaml:"ca,omitempty" json:"ca,omitempty"`
	// PEM encoded client certificate
	Cert string `yaml:"cert,omitempty" json:"cert,omitempty"`
	// PEM encoded client private key
	Key string `yaml:"key,omitempty" json:"key,omitempty"`
	InsecureSkipVerify bool
}

```

File: http/client.go [hunk 4]
```diff
@@ -175,19 +176,57 @@ func (c *Client) DisableKeepAlive(val bool) *Client {
...
+			return nil, err
+		}
+
+		if !certPool.AppendCertsFromPEM([]byte(ca)) {
+			return nil, fmt.Errorf("failed to append ca certificate")
+		}
+		customTransport.TLSClientConfig.RootCAs = certPool
+	}
+
+	if cert != "" {
```
Comment (moshloop) [inline http/client.go:207]: We need to set  here e.g.

```go
cert, _ := tls.LoadX509KeyPair("client.crt", "client.key")
customTransport.TLSClientConfig.Certificates = : []tls.Certificate{cert},
```

File: http/client.go (hunk 3 above)
Comment (moshloop) [reply http/client.go:266]: Can we also add:

TLSConfig.TLSHandshakeTimeout =  10 * time.Second,

File: http/client.go [hunk 5]
```diff
@@ -175,19 +176,71 @@ func (c *Client) DisableKeepAlive(val bool) *Client {
...
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, err
+		}
+
+		if !certPool.AppendCertsFromPEM([]byte(conf.CA)) {
+			return nil, fmt.Errorf("failed to append ca certificate")
+		}
+		transport.TLSClientConfig.RootCAs = certPool
+	} else if conf.Cert != "" && conf.Key == "" {
```
Comment (moshloop) [inline http/client.go:225]:
```suggestion
	}
	if conf.Cert != "" && conf.Key == "" {
```

File: http/client.go [hunk 6]
```diff
@@ -175,19 +176,73 @@ func (c *Client) DisableKeepAlive(val bool) *Client {
...
+			return nil, err
+		}
 
+		if !certPool.AppendCertsFromPEM([]byte(conf.CA)) {
+			return nil, fmt.Errorf("failed to append ca certificate")
+		}
+		transport.TLSClientConfig.RootCAs = certPool
+	}
+
+	if conf.Cert != "" && conf.Key == "" {
```
Comment (moshloop) [inline http/client.go:227]:
```suggestion
	if conf.Cert != "" && conf.Key != "" {
```

**Pull request merged**